### ISSUE TITLE
关于多张GPU并行计算的方法

### ISSUE THREAD

**Issue by jtyoui** [PaddlePaddle/ERNIE#97]
一张GPU显存不足，我想多张一起训练，请问在什么地方修改？？？

**Comment by SunnyGJing** [comment]
file "run_xxx.sh": 
```
# use GPU_device_1 and GPU_device_2
export CUDA_VISIBLE_DEVICES=0,1
```

**Comment by jtyoui** [comment]
> file "run_xxx.sh":
> 
> ```
> # use GPU_device_1 and GPU_device_2
> export CUDA_VISIBLE_DEVICES=0,1
> ```

Can not find library: libnccl.so. Please try to add the lib path to LD_LIBRARY_PATH？？请问这个是什么意思？我·使用两张GPU训练报的错误

**Comment by jtyoui** [comment]
Device count: 4
Num train examples: 392702
Max train steps: 18407
Num warmup steps: 0
Theoretical memory usage in training: 16075.057 - 16840.535 MB
W0416 10:10:59.425506 29150 device_context.cc:263] Please NOTE: device: 0, CUDA Capability: 35, Driver API Version: 9.1, Runtime API Version: 9.0
W0416 10:10:59.425563 29150 device_context.cc:271] device: 0, cuDNN Version: 7.0.
Load pretraining parameters from data/params.
W0416 10:11:03.200520 29150 dynamic_loader.cc:107] Can not find library: libnccl.so. Please try to add the lib path to LD_LIBRARY_PATH.
Traceback (most recent call last):
  File "run_classifier.py", line 276, in <module>
    main(args)
  File "run_classifier.py", line 178, in main
    main_program=train_program)
  File "/usr/local/lib/python3.6/site-packages/paddle/fluid/parallel_executor.py", line 191, in __init__
    local_scopes, exec_strategy, build_strategy)
paddle.fluid.core.EnforceNotMet: Failed to find dynamic library: libnccl.so ( libnccl.so: cannot open shared object file: No such file or directory ) 
 Please specify its path correctly using following ways: 
 Method. set environment variable LD_LIBRARY_PATH on Linux or DYLD_LIBRARY_PATH on Mac OS. 
 For instance, issue command: export LD_LIBRARY_PATH=... 
 Note: After Mac OS 10.11, using the DYLD_LIBRARY_PATH is impossible unless System Integrity Protection (SIP) is disabled. at [/paddle/paddle/fluid/platform/dynload/dynamic_loader.cc:163]
PaddlePaddle Call Stacks:

**Comment by jtyoui** [comment]
单张GPU训练显存不足。为什么我多张GPU一起训练，还是报显存不足？？？

**Comment by jtyoui** [comment]
> file "run_xxx.sh":
> 
> ```
> # use GPU_device_1 and GPU_device_2
> export CUDA_VISIBLE_DEVICES=0,1
> ```

请问一下：单张GPU训练显存不足。为什么我多张GPU一起训练，还是报显存不足？？？

**Comment by SunnyGJing** [comment]
> 请问一下：单张GPU训练显存不足。为什么我多张GPU一起训练，还是报显存不足？？？

把batchsize改小，可以缓解OOM问题。
我猜想你的句子长度较长，而设置的batchsize不够小。
你试一下 可以解决吗？

**Comment by AvalonSama** [comment]
> > 请问一下：单张GPU训练显存不足。为什么我多张GPU一起训练，还是报显存不足？？？
> 
> 把batchsize改小，可以缓解OOM问题。
> 我猜想你的句子长度较长，而设置的batchsize不够小。
> 你试一下 可以解决吗？

请问下在GPU上训练时，无论GPU上显存是否充足，只要有程序在跑，当前GPU是否就不可用了？
我报了
Device count: 4
Num train examples: 14788
Max train steps: 3080
Num warmup steps: 0
memory_optimize is deprecated. Use CompiledProgram and Executor
Theoretical memory usage in training: 2774.843 - 2906.979 MB
W0509 10:06:39.810094  7714 device_context.cc:261] Please NOTE: device: 0, CUDA Capability: 61, Driver API Version: 10.0, Runtime API Version: 9.0
W0509 10:06:39.813912  7714 device_context.cc:269] device: 0, cuDNN Version: 7.0.
Traceback (most recent call last):
  File "run_classifier.py", line 276, in <module>
    main(args)
  File "run_classifier.py", line 138, in main
    exe.run(startup_prog)
  File "/usr/local/anaconda3/lib/python3.6/site-packages/paddle/fluid/executor.py", line 565, in run
    use_program_cache=use_program_cache)
  File "/usr/local/anaconda3/lib/python3.6/site-packages/paddle/fluid/executor.py", line 642, in _run
    exe.run(program.desc, scope, 0, True, True, fetch_var_name)
paddle.fluid.core.EnforceNotMet: Invoke operator fill_constant error.
Python Callstacks:
  File "/usr/local/anaconda3/lib/python3.6/site-packages/paddle/fluid/framework.py", line 1725, in _prepend_op
    attrs=kwargs.get("attrs", None))
  File "/usr/local/anaconda3/lib/python3.6/site-packages/paddle/fluid/initializer.py", line 167, in __call__
    stop_gradient=True)
  File "/usr/local/anaconda3/lib/python3.6/site-packages/paddle/fluid/framework.py", line 1517, in create_var
    kwargs['initializer'](var, self)
  File "/usr/local/anaconda3/lib/python3.6/site-packages/paddle/fluid/layer_helper_base.py", line 382, in set_variable_initializer
    initializer=initializer)
  File "/usr/local/anaconda3/lib/python3.6/site-packages/paddle/fluid/layers/tensor.py", line 152, in create_global_var
    value=float(value), force_cpu=force_cpu))
  File "/usr/local/anaconda3/lib/python3.6/site-packages/paddle/fluid/optimizer.py", line 136, in _create_global_learning_rate
    persistable=True)
  File "/usr/local/anaconda3/lib/python3.6/site-packages/paddle/fluid/optimizer.py", line 275, in _create_optimization_pass
    self._create_global_learning_rate()
  File "/usr/local/anaconda3/lib/python3.6/site-packages/paddle/fluid/optimizer.py", line 441, in apply_gradients
    optimize_ops = self._create_optimization_pass(params_grads)
  File "/usr/local/anaconda3/lib/python3.6/site-packages/paddle/fluid/optimizer.py", line 469, in apply_optimize
    optimize_ops = self.apply_gradients(params_grads)
  File "/usr/local/anaconda3/lib/python3.6/site-packages/paddle/fluid/optimizer.py", line 500, in minimize
    loss, startup_program=startup_program, params_grads=params_grads)
  File "/home/kchen/ERNIE/optimization.py", line 127, in optimization
    _, param_grads = optimizer.minimize(loss)
  File "run_classifier.py", line 105, in main
    loss_scaling=args.loss_scaling)
  File "run_classifier.py", line 276, in <module>
    main(args)
C++ Callstacks:
Enforce failed. Expected allocating <= available, but received allocating:7441217234 > available:1638465280.
Insufficient GPU memory to allocation. at [/paddle/paddle/fluid/platform/gpu_info.cc:262]
PaddlePaddle Call Stacks:
0       0x7f44886bb228p void paddle::platform::EnforceNotMet::Init<std::string>(std::string, char const*, int) + 360
1       0x7f44886bb577p paddle::platform::EnforceNotMet::EnforceNotMet(std::string const&, char const*, int) + 87
2       0x7f448a403706p paddle::platform::GpuMaxChunkSize() + 630
3       0x7f448a3d80a2p
4       0x7f44b6dd3a99p
5       0x7f448a3d774dp paddle::memory::legacy::GetGPUBuddyAllocator(int) + 109
6       0x7f448a3d8573p void* paddle::memory::legacy::Alloc<paddle::platform::CUDAPlace>(paddle::platform::CUDAPlace const&, unsigned long) + 35
7       0x7f448a3d89b5p paddle::memory::allocation::LegacyAllocator::AllocateImpl(unsigned long, paddle::memory::allocation::Allocator::Attr) + 389
8       0x7f448a3fdc6bp paddle::memory::allocation::Allocator::Allocate(unsigned long, paddle::memory::allocation::Allocator::Attr) + 27
9       0x7f448a3cc5b3p paddle::memory::allocation::AllocatorFacade::Alloc(boost::variant<paddle::platform::CUDAPlace, paddle::platform::CPUPlace, paddle::platform::CUDAPinnedPlace, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&, unsigned long, paddle::memory::allocation::Allocator::Attr) + 435
10      0x7f448a3cc6d1p paddle::memory::allocation::AllocatorFacade::AllocShared(boost::variant<paddle::platform::CUDAPlace, paddle::platform::CPUPlace, paddle::platform::CUDAPinnedPlace, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&, unsigned long, paddle::memory::allocation::Allocator::Attr) + 33
11      0x7f448a00a5a0p paddle::memory::AllocShared(boost::variant<paddle::platform::CUDAPlace, paddle::platform::CPUPlace, paddle::platform::CUDAPinnedPlace, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&, unsigned long, paddle::memory::allocation::Allocator::Attr) + 48
12      0x7f448a39e96ap paddle::framework::Tensor::mutable_data(boost::variant<paddle::platform::CUDAPlace, paddle::platform::CPUPlace, paddle::platform::CUDAPinnedPlace, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_>, paddle::framework::proto::VarType_Type, paddle::memory::allocation::Allocator::Attr, unsigned long) + 154
13      0x7f44893770d1p paddle::operators::FillConstantKernel<float>::Compute(paddle::framework::ExecutionContext const&) const + 497
14      0x7f448937a273p std::_Function_handler<void (paddle::framework::ExecutionContext const&), paddle::framework::OpKernelRegistrarFunctor<paddle::platform::CUDAPlace, false, 0ul, paddle::operators::FillConstantKernel<float>, paddle::operators::FillConstantKernel<double>, paddle::operators::FillConstantKernel<long>, paddle::operators::FillConstantKernel<int>, paddle::operators::FillConstantKernel<paddle::platform::float16> >::operator()(char const*, char const*, int) const::{lambda(paddle::framework::ExecutionContext const&)#1}>::_M_invoke(std::_Any_data const&, paddle::framework::ExecutionContext const&) + 35
15      0x7f448a34a376p paddle::framework::OperatorWithKernel::RunImpl(paddle::framework::Scope const&, boost::variant<paddle::platform::CUDAPlace, paddle::platform::CPUPlace, paddle::platform::CUDAPinnedPlace, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&, paddle::framework::RuntimeContext*) const + 662
16      0x7f448a34aae4p paddle::framework::OperatorWithKernel::RunImpl(paddle::framework::Scope const&, boost::variant<paddle::platform::CUDAPlace, paddle::platform::CPUPlace, paddle::platform::CUDAPinnedPlace, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&) const + 292
17      0x7f448a34840cp paddle::framework::OperatorBase::Run(paddle::framework::Scope const&, boost::variant<paddle::platform::CUDAPlace, paddle::platform::CPUPlace, paddle::platform::CUDAPinnedPlace, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&) + 332
18      0x7f448882d3fep paddle::framework::Executor::RunPreparedContext(paddle::framework::ExecutorPrepareContext*, paddle::framework::Scope*, bool, bool, bool) + 382
19      0x7f448882e23fp paddle::framework::Executor::Run(paddle::framework::ProgramDesc const&, paddle::framework::Scope*, int, bool, bool, std::vector<std::string, std::allocator<std::string> > const&, bool) + 143
20      0x7f44886aa8dep
21      0x7f44886ed7cep
22      0x5599bc7631a4p _PyCFunction_FastCallDict + 340
23      0x5599bc7ea58ep
24      0x5599bc80ea1ap _PyEval_EvalFrameDefault + 762
25      0x5599bc7e3814p
26      0x5599bc7e46b1p
27      0x5599bc7ea515p
28      0x5599bc80f734p _PyEval_EvalFrameDefault + 4116
29      0x5599bc7e3814p
30      0x5599bc7e46b1p
31      0x5599bc7ea515p
32      0x5599bc80ea1ap _PyEval_EvalFrameDefault + 762
33      0x5599bc7e447bp
34      0x5599bc7ea515p
35      0x5599bc80ea1ap _PyEval_EvalFrameDefault + 762
36      0x5599bc7e51c9p PyEval_EvalCodeEx + 809
37      0x5599bc7e5f5cp PyEval_EvalCode + 28
38      0x5599bc8685e4p
39      0x5599bc8689e1p PyRun_FileExFlags + 161
40      0x5599bc868be4p PyRun_SimpleFileExFlags + 452
41      0x5599bc86c6c4p Py_Main + 1572
42      0x5599bc73452ep main + 238
43      0x7f44b6a1b830p __libc_start_main + 240
44      0x5599bc81557fp
这四张卡上都有程序再跑，但是并没有完全占用显存。
